# Проект по SQL

**Описание проекта**: Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

**Задачи**: 
1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.
 
**Цель исследования**: Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица** `books`

Содержит данные о книгах:

 - `book_id` — идентификатор книги;
 - `author_id` — идентификатор автора;
 - `title` — название книги;
 - `num_pages` — количество страниц;
 - `publication_date` — дата публикации книги;
 - `publisher_id` — идентификатор издателя.
 
**Таблица** `authors`
 
Содержит данные об авторах:

 - `author_id` — идентификатор автора;
 - `author` — имя автора.

**Таблица** `publishers`

Содержит данные об издательствах:

 - `publisher_id` — идентификатор издательства;
 - `publisher` — название издательства;

**Таблица** `ratings`

Содержит данные о пользовательских оценках книг:

 - `rating_id` — идентификатор оценки;
 - `book_id` — идентификатор книги;
 - `username` — имя пользователя, оставившего оценку;
 - `rating` — оценка книги.

**Таблица** `reviews`

Содержит данные о пользовательских обзорах:

 - `review_id` — идентификатор обзора;
 - `book_id` — идентификатор книги;
 - `username` — имя автора обзора;
 - `text` — текст обзора.
 
## Схема данных


![подпись](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * 
           FROM books 
           LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query = '''SELECT * 
           FROM authors 
           LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query = '''SELECT * 
           FROM publishers 
           LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
query = '''SELECT * 
           FROM ratings 
           LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
query = '''SELECT * 
           FROM reviews 
           LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Данные из таблиц, соответствуют описанию и обозначенной структуре.

**Посчитайте, сколько книг вышло после 1 января 2000 года**

In [9]:
query = '''SELECT COUNT(book_id) 
           FROM books
           WHERE publication_date > '2000-01-01' '''

pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2000 года вышло 819 книг.

**Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [10]:
query = '''SELECT b.book_id,
                  b.title,
                  COUNT(DISTINCT re.review_id) AS num_reviews,
                  AVG(ra.rating) AS avg_rating
           FROM books AS b
           LEFT JOIN reviews AS re ON b.book_id = re.book_id
           JOIN ratings AS ra ON b.book_id = ra.book_id
           GROUP BY b.book_id'''

pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  num_reviews  \
0          1                                       'Salem's Lot            2   
1          2                 1 000 Places to See Before You Die            1   
2          3  13 Little Blue Envelopes (Little Blue Envelope...            3   
3          4  1491: New Revelations of the Americas Before C...            2   
4          5                                               1776            4   
..       ...                                                ...          ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)            3   
996      997                        Xenocide (Ender's Saga  #3)            3   
997      998                                    Year of Wonders            4   
998      999                        You Suck (A Love Story  #2)            2   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...            4   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 4 columns]

Посчитано количество обзоров и средняя оценка по каждой из книг.

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [11]:
query = '''SELECT p.publisher,
                  COUNT(b.book_id)
           FROM publishers AS p
           JOIN books AS b ON b.publisher_id = p.publisher_id
           WHERE b.num_pages > 50
           GROUP BY p.publisher
           ORDER BY COUNT(b.book_id) DESC
           LIMIT 1'''

pd.io.sql.read_sql(sql=text(query), con = con)

publisher  count
0  Penguin Books     42

Penguin Books является издательством, выпустившее наибольшее число книг толще 50 страниц.

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [12]:
query = '''SELECT a.author,
                  AVG(r.rating)
           FROM authors AS a
           JOIN (SELECT b.author_id, b.book_id
                 FROM books AS b
                 JOIN ratings AS r ON b.book_id = r.book_id
                 GROUP BY b.book_id
                 HAVING COUNT(r.rating_id) >= 50) AS filtered_books ON a.author_id = filtered_books.author_id
           JOIN ratings AS r ON filtered_books.book_id = r.book_id
           GROUP BY a.author
           HAVING COUNT(r.rating_id) >= 50
           ORDER BY AVG(r.rating) DESC
           LIMIT 1'''

pd.io.sql.read_sql(sql=text(query), con = con)

author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

J.K. Rowling/Mary GrandPré является автором с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [13]:
query = '''SELECT AVG(total_reviews)
           FROM (SELECT username,
                        COUNT(rating_id) AS total_ratings
                 FROM ratings
                 GROUP BY username
                 HAVING COUNT(rating_id) > 48) AS rated_users
           JOIN (SELECT username,
                        COUNT(review_id) AS total_reviews
                 FROM reviews
                 GROUP BY username) AS reviewed_users ON rated_users.username = reviewed_users.username'''

pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

Среди пользователей, которые поставили больше 48 оценок, среднее количество обзоров составляет 24.

## Общий вывод

1. Данные из таблиц, соответствуют описанию и обозначенной структуре.
2. После 1 января 2000 года вышло 819 книг.
3. Посчитано количество обзоров и средняя оценка по каждой из книг.
4. Penguin Books является издательством, выпустившее наибольшее число книг толще 50 страниц.
5. J.K. Rowling/Mary GrandPré является автором с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.
6. Среди пользователей, которые поставили больше 48 оценок, среднее количество обзоров составляет 24.

**Рекомендации**: Заказчику рекомендуется сосредоточить внимание на книгах, изданным Penguin Books. Также рекомендуется рассмотреть сотрудничество с автором J.K. Rowling/Mary GrandPré, чьи книги получили высокие оценки читателей.